### PR TITLE
Document and check for unsupported Ruby versions

### PR DESCRIPTION
**Pull request by dazuma**
I added code that will check for the current `RUBY_VERSION` at `google-cloud-core` load time, and print warnings if an old version is detected.

Also proposed some language in the README (just edited google-cloud-core's readme for now). If this looks good, I'll replicate it over the other readmes.

### CODE REVIEW

File: google-cloud-core/README.md [hunk 1]
```diff
@@ -6,7 +6,14 @@ This library contains shared types, such as error classes, for the google-cloud
 
 ## Supported Ruby Versions
 
-This library is supported on Ruby 2.0+.
+This library is currently supported on Ruby 2.0+.
+
+However, Ruby 2.3 or later is strongly recommended, as earlier releases are
```
Comment (quartzmo) [inline google-cloud-core/README.md]: "are or are nearing end-of-life" -> "are at or are nearing end-of-life" or "have reached or are nearing end-of-life"?

File: google-cloud-core/README.md [hunk 2]
```diff
@@ -6,7 +6,14 @@ This library contains shared types, such as error classes, for the google-cloud
...
+or are nearing end-of-life. After June 1, 2018, Google will provide official
+support only for Ruby versions that are considered current and supported by
+Ruby-core (that is, Ruby versions that are either in normal maintenance or in
```
Comment (quartzmo) [inline google-cloud-core/README.md]: "Ruby-core" -> "Ruby Core" (Or is it "Ruby-Core"?)

Comment (dazuma) [reply google-cloud-core/README.md]: That's a good question. I'm not sure there's a real "official" entity here. I think people generally refer to it as the "Ruby core team"?

Comment (quartzmo) [reply google-cloud-core/README.md]: https://www.ruby-lang.org/en/community/ruby-core/ is titled "Ruby Core" but the [mailing list](https://www.ruby-lang.org/en/community/mailing-lists/) is "Ruby-Core". Beats me, but I think capitalizing "Core" is ok.

File: google-cloud-core/lib/google/cloud.rb [hunk 3]
```diff
@@ -96,26 +96,81 @@ def self.configure
...
+    # Check Ruby version and emit a warning if it is old
+    # @private
+    #
+    def self.warn_on_old_ruby_version \
+        supported_version: SUPPORTED_VERSION_THRESHOLD,
+        recommended_version: RECOMMENDED_VERSION_THRESHOLD
+      cur_version = Gem::Version.new RUBY_VERSION
+      if cur_version < Gem::Version.new(supported_version)
+        warn "WARNING: You are running Ruby #{cur_version}, which has reached" \
+          " end-of-life and is no longer supported by Ruby-core."
```
Comment (quartzmo) [inline google-cloud-core/lib/google/cloud.rb]: Change "Ruby-core" here too if desired.

File: google-cloud-core/README.md (hunk 1 above)
Comment (dazuma) [reply google-cloud-core/README.md]: changed to "have reached or are nearing end-of-life"

File: google-cloud-core/README.md (hunk 2 above)
Comment (dazuma) [reply google-cloud-core/README.md]: Changed to "Ruby Core"

File: google-cloud-core/lib/google/cloud.rb (hunk 3 above)
Comment (dazuma) [reply google-cloud-core/lib/google/cloud.rb]: Changed to "Ruby Core"

**Comment by blowmage** [comment]
When do you want this warning to be released?

**Comment by blowmage** [comment]
@dazuma Looking at the build output, it isn't obvious what library the warning is coming from. Should we make it more obvious that the warning is coming from "Google Cloud"?

**Comment by dazuma** [comment]
@blowmage That's a good idea. I'll update the wording. And let's release a google-cloud-core update maybe late this week if that works for you.

**Comment by dazuma** [comment]
@blowmage Also what do you think of providing an environment variable or something that users can set to suppress warnings like this (and maybe the config validation warnings)?

**Comment by blowmage** [comment]
I think giving users a mechanism to suppress the warning is a good idea. I don't have a preference on the mechanism, but an environment variable seems like a logical choice.

**Pull request merged**